staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -293,9 +377,11 @@ type Cacher struct {
 	// watchersToStop is a list of watchers that were supposed to be stopped
 	// during current dispatching, but stopping was deferred to the end of
 	// dispatching that event to avoid race with closing channels in watchers.
-	watchersToStop []*cacheWatcher
+	watchersToStop []*watchersToStopWithDrainTuple
 	// Maintain a timeout queue to send the bookmark event before the watcher times out.
 	bookmarkWatchers *watcherBookmarkTimeBuckets
+
+	bookmarkAfterRVWatchers *bookmarkAfterResourceVersionWatchers
```

Comment:
Please correct me if I'm missing something, but I'm not sure I like the current approach.
This is introducing a bunch of new code and fields and I'm not convinced we really need that (vs putting this functionality elsewhere).

1. We need a way to distinguish a watch that should be running "until a given RV" from the regular ones.
I would much rather prefer to put that login into `cacheWatcher` directly and let it handle everything inside.
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L1148
[cacheWatcher is much simpler thing to follow then the overall Cacher and it would simplify the reasoning, debugging, etc.]

2. The way I'm currently thinking about it is:
- cacheWatcher gets an additional parameter of "endRV' (better name needed)
- the overall Cacher remains unchanged
- inside cacheWatcher we add the logic of proper handling termination, so at the high level (I'm sure things are missing here)
  (a) when it gets bookmark and isn't yet synced, just ignore it
  (b) if it gets event with RV higher than its EndRV, just ignore it a stop itself

[With the above we're technically risking having some already stopped watchers in bookmarkWatchers, but that's fine - this is possible even now - they are just GCed when they bookmark time approaches.]

Please take a look and let me know what do you think about it.